src/riscv/lib/src/jit/builder/instruction.rs

```diff
@@ -0,0 +1,577 @@
+    fn create_unknown_branch_outcome(&mut self, destination: Value) -> Block {
+        let hook = self.builder.create_block();
+        self.outcomes
+            .push(Outcome::UnknownBranch { destination, hook });
+        hook
+    }
+
+    /// Handle an exception raised by the instruction.
+    fn handle_exception<Any>(&mut self, exception_ptr: Value) -> InstructionResult<Any>
+    where
+        M: JitStateAccess,
+    {
+        let current_pc = self.pc_read();
+        let outcome = self.ext_calls.handle_exception(
+            self.builder,
+            self.core_param,
+            exception_ptr,
+            self.result_param,
+            current_pc,
+        );
+
+        let exception_block = self.create_exception_outcome();
+        let unknown_branch_block = self.create_unknown_branch_outcome(outcome.new_pc.0);
+
+        self.ins().brif(
+            outcome.handled,
+            unknown_branch_block,
+            [],
+            exception_block,
+            [],
+        );
+
+        // The predecessors of either block are now known
+        self.builder.seal_block(exception_block);
+        self.builder.seal_block(unknown_branch_block);
+
+        InstructionResult::NoNext
+    }
+
+    /// Finalise the instruction building and produce an instruction.
+    pub fn finish(
+        self,
+        result: InstructionResult<ProgramCounterUpdate<X64>>,
+    ) -> LoweredInstruction {
+        let mut lowered = LoweredInstruction {
+            program_counter: self.instruction_pc,
+            run_block: self.entry_block,
+            outcomes: self.outcomes,
+        };
+
+        // Hook up the end of the instruction.
+        match result {
+            InstructionResult::NoNext => {
+                // When the instruction being built exits regardless, that means that the block
+                // we're currently targeting ends in a branching or jump IR instruction.
+            }
+
+            InstructionResult::HasNext(update) => {
+                // However, when a next instruction is possible, the current block needs to be
+                // populated. In this case, we jump to the corresponding outcome hook block. We
+                // need to insert this jump instruction to ensure that the block is not empty -
+                // otherwise we can't switch away from it.
+
+                let hook = self.builder.create_block();
+                self.builder.ins().jump(hook, []);
+
+                let outcome = match update {
+                    ProgramCounterUpdate::Set(address) => Outcome::UnknownBranch {
+                        destination: address.0,
+                        hook,
+                    },
+                    ProgramCounterUpdate::Next(_width) => Outcome::Next { hook },
+                };
+                lowered.outcomes.push(outcome);
+            }
+        }
+
+        lowered
+    }
+}
+
+impl<MC: MemoryConfig, M: JitStateAccess> ICB for InstructionBuilder<'_, '_, MC, M> {
```

Comment:
IIUC the plan to start augmenting all the functions from this impl with cranelift IR which will track program counter updates. I noticed that a lot of these functions can actually be implemented directly on top of `self.builder: FunctionBuilder`. My suggestion is that at least for the "simple" instructions that do not branch / jump, the flow can be:

* Generate IR from calling `self.builder.<this_operation>(...)`
* Add IR for Incrementing program counter

For the more involved cases remains to be seen how it will turn out, but maybe there is some opportunity for those to separate concerns as well

Reply:
I'm not sure I understand what you mean. This is the builder for a single instruction; the program counter is not incremented during an instruction.

Reply:
Oh, you're right, the IR for program counter is added in the `SequenceBuilder`